dev/TEAL_opcodes.md

```diff
@@ -0,0 +1,447 @@
+
+`bytecblock` loads the following program bytes into an array of byte string constants in the evaluator. These constants can be referred to by `bytec` and `bytec_*` which will push the value onto the stack.
+
+## bytec
+
+- Opcode: 0x27 {uint8 byte constant index}
+- Pops: _None_
+- Pushes: []byte
+- push bytes constant to stack by index into constants
+
+## bytec_0
+
+- Opcode: 0x28 
+- Pops: _None_
+- Pushes: []byte
+- push bytes constant 0 to stack
+
+## bytec_1
+
+- Opcode: 0x29 
+- Pops: _None_
+- Pushes: []byte
+- push bytes constant 1 to stack
+
+## bytec_2
+
+- Opcode: 0x2a 
+- Pops: _None_
+- Pushes: []byte
+- push bytes constant 2 to stack
+
+## bytec_3
+
+- Opcode: 0x2b 
+- Pops: _None_
+- Pushes: []byte
+- push bytes constant 3 to stack
+
+## arg
+
+- Opcode: 0x2c {uint8 arg index N}
+- Pops: _None_
+- Pushes: []byte
+- push LogicSig.Args[N] value to stack by index
+
+## arg_0
+
+- Opcode: 0x2d 
+- Pops: _None_
+- Pushes: []byte
+- push LogicSig.Args[0] to stack
+
+## arg_1
+
+- Opcode: 0x2e 
+- Pops: _None_
+- Pushes: []byte
+- push LogicSig.Args[1] to stack
+
+## arg_2
+
+- Opcode: 0x2f 
+- Pops: _None_
+- Pushes: []byte
+- push LogicSig.Args[2] to stack
+
+## arg_3
+
+- Opcode: 0x30 
+- Pops: _None_
+- Pushes: []byte
+- push LogicSig.Args[3] to stack
+
+## txn
+
+- Opcode: 0x31 {uint8 transaction field index}
+- Pops: _None_
+- Pushes: any
+- push field from current transaction to stack
+
+`txn` Fields:
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | Sender | []byte |
+| 1 | Fee | uint64 |
+| 2 | FirstValid | uint64 |
+| 3 | FirstValidTime | uint64 |
+| 4 | LastValid | uint64 |
+| 5 | Note | []byte |
+| 6 | Lease | []byte |
+| 7 | Receiver | []byte |
+| 8 | Amount | uint64 |
+| 9 | CloseRemainderTo | []byte |
+| 10 | VotePK | []byte |
+| 11 | SelectionPK | []byte |
+| 12 | VoteFirst | uint64 |
+| 13 | VoteLast | uint64 |
+| 14 | VoteKeyDilution | uint64 |
+| 15 | Type | []byte |
+| 16 | TypeEnum | uint64 |
+| 17 | XferAsset | uint64 |
+| 18 | AssetAmount | uint64 |
+| 19 | AssetSender | []byte |
+| 20 | AssetReceiver | []byte |
+| 21 | AssetCloseTo | []byte |
+| 22 | GroupIndex | uint64 |
+| 23 | TxID | []byte |
+
+
+TypeEnum mapping:
+
+| Index | "Type" string | Description |
+| --- | --- | --- |
+| 0 | unknown | Unknown type. Invalid transaction. |
+| 1 | pay | Payment |
+| 2 | keyreg | Key Registration |
+| 3 | acfg | Asset Config |
+| 4 | axfer | Asset Transfer |
+| 5 | afrz | Asset Freeze |
+
+
+FirstValidTime is actually the time of the round at FirstValid-1. Subtle implementation details make it much faster to serve details of an already completed round.
+
+## global
+
+- Opcode: 0x32 {uint8 global field index}
+- Pops: _None_
+- Pushes: any
+- push value from globals to stack
+
+`global` Fields:
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | MinTxnFee | uint64 |
+| 1 | MinBalance | uint64 |
+| 2 | MaxTxnLife | uint64 |
+| 3 | ZeroAddress | []byte |
+| 4 | GroupSize | uint64 |
+
+
+## gtxn
```

Comment:
Add that for a transaction with group index i, `gtxn i fieldIndex` has the same behaviour as `txn fieldIndex`.